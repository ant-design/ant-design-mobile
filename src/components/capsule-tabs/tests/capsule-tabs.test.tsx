import React, { useState } from 'react'
import { fireEvent, render, testA11y } from 'testing'
import CapsuleTabs, { CapsuleTabsProps } from '..'
import ConfigProvider from '../../config-provider'

const classPrefix = `adm-capsule-tabs`

describe('CapsuleTabs', () => {
  const Basic = (props: CapsuleTabsProps) => (
    <CapsuleTabs {...props}>
      <CapsuleTabs.Tab title='fruits' key='fruits'>
        Apple
      </CapsuleTabs.Tab>
      <CapsuleTabs.Tab title='vegetables' key='vegetables'>
        Tomato
      </CapsuleTabs.Tab>
      <CapsuleTabs.Tab title='animals' key='animals'>
        Ant
      </CapsuleTabs.Tab>
    </CapsuleTabs>
  )
  test('a11y', async () => {
    await testA11y(<Basic />)
  })

  test('basic usage', async () => {
    const { getByText } = render(<Basic defaultActiveKey='animals' />)
    expect(getByText('Ant')).toBeVisible()
    expect(getByText('animals')).toHaveClass(`${classPrefix}-tab-active`)
    fireEvent.click(getByText('vegetables'))
    expect(getByText('Tomato')).toBeVisible()
    expect(getByText('vegetables')).toHaveClass(`${classPrefix}-tab-active`)
  })

  test('controlled mode', async () => {
    const App = () => {
      const [activeKey, setActiveKey] = useState<string | null>(null)
      return <Basic activeKey={activeKey} onChange={key => setActiveKey(key)} />
    }
    const { getByText } = render(<App />)
    expect(document.querySelectorAll(`.${classPrefix}-tab-active`).length).toBe(
      0
    )
    fireEvent.click(getByText('vegetables'))
    expect(getByText('Tomato')).toBeVisible()
  })

  test('disabled tab', async () => {
    const onChange = jest.fn()
    const { getByText } = render(
      <CapsuleTabs onChange={onChange}>
        <CapsuleTabs.Tab title='fruits' key='fruits' />
        <CapsuleTabs.Tab title='vegetables' key='vegetables' />
        <CapsuleTabs.Tab title='animals' key='animals' disabled />
      </CapsuleTabs>
    )

    expect(getByText('animals')).toHaveClass(`${classPrefix}-tab-disabled`)
    fireEvent.click(getByText('animals'))
    expect(onChange).not.toBeCalled()
  })

  test('render the DOM structure when hidden', async () => {
    const { queryByText } = render(
      <CapsuleTabs>
        <CapsuleTabs.Tab title='fruits' key='fruits'>
          Apple
        </CapsuleTabs.Tab>
        <CapsuleTabs.Tab title='vegetables' key='vegetables'>
          Tomato
        </CapsuleTabs.Tab>
        <CapsuleTabs.Tab title='animals' key='animals' forceRender>
          Ant
        </CapsuleTabs.Tab>
      </CapsuleTabs>
    )

    expect(queryByText('Ant')).toBeInTheDocument()
  })

  test('unmount content when not visible', async () => {
    const { getByText, queryByText } = render(
      <CapsuleTabs>
        <CapsuleTabs.Tab title='fruits' key='fruits' destroyOnClose>
          Apple
        </CapsuleTabs.Tab>
        <CapsuleTabs.Tab title='vegetables' key='vegetables'>
          Tomato
        </CapsuleTabs.Tab>
        <CapsuleTabs.Tab title='animals' key='animals'>
          Ant
        </CapsuleTabs.Tab>
      </CapsuleTabs>
    )
    expect(queryByText('Apple')).toBeInTheDocument()
    fireEvent.click(getByText('vegetables'))
    expect(queryByText('Apple')).not.toBeInTheDocument()
  })

  test('should apply prefixCls from ConfigProvider', () => {
    const { container } = render(
      <ConfigProvider prefixCls='config-prefix'>
        <CapsuleTabs>
          <CapsuleTabs.Tab title='fruits' key='fruits' />
        </CapsuleTabs>
      </ConfigProvider>
    )
    expect(container.querySelector('.config-prefix-capsule-tabs')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  test('should prioritize component prefixCls over ConfigProvider', () => {
    const { container } = render(
      <ConfigProvider prefixCls='config-prefix'>
        <CapsuleTabs prefixCls='component-prefix'>
          <CapsuleTabs.Tab title='fruits' key='fruits' />
        </CapsuleTabs>
      </ConfigProvider>
    )
    expect(container.querySelector('.component-prefix')).toBeTruthy()
    expect(container.querySelector('.config-prefix-capsule-tabs')).toBeFalsy()
    expect(container).toMatchSnapshot()
  })
})
