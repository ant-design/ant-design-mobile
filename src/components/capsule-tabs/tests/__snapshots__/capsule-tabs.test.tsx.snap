// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CapsuleTabs should apply prefixCls from ConfigProvider 1`] = `
<div>
  <div
    class="config-prefix-capsule-tabs"
  >
    <div
      class="config-prefix-capsule-tabs-header"
    >
      <div
        class="adm-scroll-mask adm-scroll-mask-left"
        style="opacity: 0;"
      />
      <div
        class="adm-scroll-mask adm-scroll-mask-right"
        style="opacity: 0;"
      />
      <div
        class="config-prefix-capsule-tabs-tab-list"
      >
        <div
          class="config-prefix-capsule-tabs-tab-wrapper"
        >
          <div
            class="config-prefix-capsule-tabs-tab config-prefix-capsule-tabs-tab-active"
          >
            fruits
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CapsuleTabs should prioritize component prefixCls over ConfigProvider 1`] = `
<div>
  <div
    class="component-prefix"
  >
    <div
      class="component-prefix-header"
    >
      <div
        class="adm-scroll-mask adm-scroll-mask-left"
        style="opacity: 0;"
      />
      <div
        class="adm-scroll-mask adm-scroll-mask-right"
        style="opacity: 0;"
      />
      <div
        class="component-prefix-tab-list"
      >
        <div
          class="component-prefix-tab-wrapper"
        >
          <div
            class="component-prefix-tab component-prefix-tab-active"
          >
            fruits
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
